### PR TITLE
feat(planner): add Flow Matching trajectory decoder as swappable planner

### LINKED ISSUE #45: Proposal: Diffusion decoder with flow matching as optional trajectory planner
## Summary

Unlike discrete textual tokens or autoregressive sequence decoding (which can result in jerky, computationally inefficient trajectory predictions), flow matching operates over continuous values directly producing a smooth continuous motion output - as implemented in the Alpamayo model. This proposal seeks to add a swappable module in the trajectory planner to use flow matching methodology.

## Key Benefits

- Continuous Trajectories: It defines a probability path (a velocity field) connecting noise to the final trajectory plan.
- Geometry-Aware Control: By bypassing discrete motion tokens, the decoder preserves scalar geometry and kinematic constraints natively.
- Real-time Latency: The optimization streamlines both training and inference, achieving single-digit millisecond inference times necessary for safety
- Continuous Motion Output: The flow-matching head regresses the velocity vectors to map the straightest possible path from the starting state to the optimal, multi-modal future trajectory.

## Implementation Proposal
To integrate a flow-matching based trajectory decoder into a Bird's-Eye-View (BEV) feature module, we couldt transform the 2D spatially grid-aligned BEV features into an information-dense conditional context vector. This vector would then guide a continuous Vector Field (Velocity Field) Regressor to denoise raw gaussian noise into smooth, kinematically sound coordinates.

### Architectural Changes
```
BEV Feature Map] ────> [Conv2D / Flatten] ────┐
(B, C, H, W)                                  │
                                              ▼
[Noisy Actions u_t] ──> [Action Projection] ──> [Cross-Attention Layer] ──> [AdaLN (Time t)] ──> [Velocity Field Output v_θ]
(B, T, 2)                                                                                       (B, T, 2)

```
```python
class TimestepEmbedding(nn.Module):
    """Maps scalar timesteps t to high-dimensional sinusoidal embeddings."""
    def __init__(self, dim):
        super().__init__()
        self.dim = dim

    def forward(self, t):
        device = t.device
        half_dim = self.dim // 2
        embeddings = math.log(10000) / (half_dim - 1)
        embeddings = torch.exp(torch.arange(half_dim, device=device) * -embeddings)
        embeddings = t[:, None] * embeddings[None, :]
        embeddings = torch.cat((embeddings.sin(), embeddings.cos()), dim=-1)
        return embeddings

class FlowMatchingControlDecoder(nn.Module):
    def __init__(self, bev_channels=256, embed_dim=256, traj_len=64, num_heads=4):
        super().__init__()
        self.traj_len = traj_len
        self.embed_dim = embed_dim
        
        # Time Step Embedding Generator
        self.time_mlp = nn.Sequential(
            TimestepEmbedding(embed_dim),
            nn.Linear(embed_dim, embed_dim),
            nn.SiLU(),
            nn.Linear(embed_dim, embed_dim)
        )
        
        # Action Input Projection (Maps continuous [acceleration, curvature] pairs)
        self.action_proj = nn.Linear(2, embed_dim)
        
        # BEV Image Feature Processing 
        # Compresses the spatial map or downsamples if map resolution is massive
        self.bev_downsample = nn.Sequential(
            nn.Conv2d(bev_channels, embed_dim, kernel_size=3, stride=1, padding=1),
            nn.GroupNorm(32, embed_dim),
            nn.SiLU()
        )
        
        # Attention Layer: Action queries look over the entire spatial BEV image map
        self.cross_attn = nn.MultiheadAttention(embed_dim, num_heads, batch_first=True)
        
        # Layer Normalization Blocks
        self.norm_bev = nn.LayerNorm(embed_dim)
        self.norm_action = nn.LayerNorm(embed_dim)
        
        # Adaptive Layer Normalization (AdaLN) for condition scaling via time injection
        self.adaln_gamma = nn.Linear(embed_dim, embed_dim)
        self.adaln_beta = nn.Linear(embed_dim, embed_dim)
        
        # Prediction Output Head: Regresses the velocity field vectors (da/dt, d_kappa/dt)
        self.velocity_head = nn.Sequential(
            nn.Linear(embed_dim, embed_dim),
            nn.SiLU(),
            nn.Linear(embed_dim, 2)
        )

    def forward(self, u_t, t, bev_features):
        """
        u_t: Noisy control actions at flow time step t. Shape: (B, Traj_Len, 2) -> [accel, curvature]
        t: Continuous flow-matching time step. Shape: (B,) scalar tensor between [0.0, 1.0]
        bev_features: Image feature tensor from your BEV module backbone. Shape: (B, C, H, W)
        """
        B, _, H, W = bev_features.shape
        
        # 1. Process and encode continuous flow time
        t_embed = self.time_mlp(t) # Shape: (B, embed_dim)
        
        # 2. Flatten 2D BEV features into an attention-ready sequence
        processed_bev = self.bev_downsample(bev_features) # Shape: (B, embed_dim, H, W)
        bev_flat = processed_bev.permute(0, 2, 3, 1).reshape(B, H * W, self.embed_dim)
        bev_context = self.norm_bev(bev_flat) # Shape: (B, H*W, embed_dim)
        
        # 3. Project continuous action trajectories to Query space
        action_query = self.norm_action(self.action_proj(u_t)) # Shape: (B, Traj_Len, embed_dim)
        
        # 4. Perform cross-attention feature fusion
        # Queries (Actions) search through Keys & Values (BEV Context)
        attn_out, _ = self.cross_attn(query=action_query, key=bev_context, value=bev_context)
        
        # 5. Inject time embeddings using AdaLN modulation
        gamma = self.adaln_gamma(t_embed).unsqueeze(1) # Shape: (B, 1, embed_dim)
        beta = self.adaln_beta(t_embed).unsqueeze(1)   # Shape: (B, 1, embed_dim)
        fused_features = attn_out * (1 + gamma) + beta
        
        # 6. Regress Vector Field Velocity
        v_theta = self.velocity_head(fused_features) # Shape: (B, Traj_Len, 2)
        return v_theta

### PR DESCRIPTION
**Pull request by riita10069**
## Summary

Implements #45 — adds a Flow Matching trajectory decoder as a swappable alternative to the GRU planner, via a registry pattern (mirroring view fusion).

## Architecture

```
Model/model_components/trajectory_planning/
├── __init__.py          ← PLANNER_REGISTRY + build_planner()
├── base.py              ← BasePlanner interface
├── gru_planner.py       ← existing autoregressive GRU planner (moved here)
└── flow_matching_planner.py  ← new Flow Matching decoder
```

`AutoE2E(planner_mode="gru" | "flow_matching")` selects the decoder. Both produce `(trajectory, ego_hidden)`.

## Flow Matching Planner

Following Lipman et al. (2023), trains a conditional velocity field `v_theta(u_t, t, c)` mapping noise `x_0 ~ N(0,I)` to the target trajectory along the linear OT path `u_t = (1-t)·x_0 + t·x_1`, with target velocity `x_1 - x_0`.

**Spatial structure preserved via cross-attention:** the noisy trajectory is a sequence of `num_timesteps` action tokens that query the flattened BEV spatial map (H·W keys/values) through multi-head cross-attention. Time + ego/visual_history conditioning are injected via AdaLN affine modulation (the DiT pattern). This avoids collapsing the BEV grid — important now that BEV is 450×300.

**Training/inference API (Alpamayo-style external loss):**
```python
u_t, t, target_velocity = planner.construct_training_data(trajectory_target)
v_pred, ego_hidden = model(x, vis_hist, ego, mode="train",
                           noisy_trajectory=u_t, flow_timestep=t)
loss = F.mse_loss(v_pred, target_velocity)
```
Inference Euler-integrates `dx/dt = v_theta` over `num_inference_steps` (default 10), with an optional `generator` for reproducibility. The projected BEV key/value sequence is computed once and reused across ODE steps.

## Test plan

- [x] 152 unit tests pass (GRU + flow matching + registry)
- [x] construct_training_data → forward → MSE loss → backward verified end-to-end
- [x] Reproducibility: same generator seed yields identical inference trajectories
- [x] Velocity field depends on BEV, time, and conditioning
- [x] GRU planner behavior unchanged after move (backward-compat tests)

## Refs

- Closes #45

### CODE REVIEW

**Review by m-zain-khawaja — APPROVED** [review]
Approved - thanks @riita10069
